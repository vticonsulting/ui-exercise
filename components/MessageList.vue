<template>
  <table class="w-full">
    <thead v-if="messages.length">
      <tr class="flex border-b text-2xl">
        <th class="pr-3 md:pl-3 text-left font-normal">
          <input
            class="form-checkbox"
            type="checkbox"
            @click="checkAll()"
            v-model="isCheckAll"
          />
        </th>
        <th class="flex items-center text-left font-normal pr-2 flex-grow">
          <div :class="[checkedMessages.length ? 'hidden' : 'flex']">
            <a class="mr-1" href="#" title="Refresh">
              <RefreshIcon class="w-5 h-5" />
            </a>
          </div>
          <div
            v-show="checkedMessages.length"
            class="flex items-center space-x-2"
          >
            <a href="#" title="Archive selected">
              <ArchiveIcon class="w-5 h-5" />
            </a>
            <a href="#" title="Report spam">
              <AlertIcon class="w-5 h-5" />
            </a>
            <a title="Delete">
              <DeleteIcon class="w-5 h-5" />
            </a>
            <a href="#" title="Mark as read">
              <MailIcon class="w-5 h-5" />
            </a>
            <a href="#" title="Snooze">
              <ClockIcon class="w-5 h-5" />
            </a>
          </div>
          <a class="mr-1" href="#" title="More">
            <MoreVerticalIcon class="w-5 h-5" />
          </a>
        </th>
        <th class="msg__col text-left font-normal">
          <a href="#" title="Settings">
            <SettingsIcon class="w-5 h-5" />
          </a>
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        v-for="message in messages"
        :key="message.id"
        class="relative group hover:bg-gray-200 hover:shadow cursor-pointer flex flex-col md:flex-row border-b"
      >
        <td class="md:py-2 md:px-3">
          <input
            class="form-checkbox"
            type="checkbox"
            :value="message"
            v-model="checkedMessages"
            @change="updateCheckAll()"
          />
        </td>
        <td class="md:py-2 pr-6" @click.stop="goTo(message.id)">
          {{ message.sender }}
        </td>
        <td
          class="md:py-2 pr-1 text-black whitespace-no-wrap"
          v-if="message.tags.length"
          @click.stop="goTo(message.id)"
        >
          <span
            class="mr-1 p-1 bg-gray-400 rounded text-xs capitalize"
            :key="tag"
            v-for="tag in message.tags"
            :class="tag"
          >
            {{ tag }}
          </span>
        </td>
        <td
          class="md:py-2 pr-4 text-black font-medium xl:flex-no-grow"
          @click.stop="goTo(message.id)"
        >
          {{ message.subject }}
        </td>
        <td
          class="md:py-2 hidden xl:block pr-2 flex-grow"
          @click.stop="goTo(message.id)"
        >
          {{ truncate(message.body) }}
        </td>
        <td
          class="md:py-2 pr-2 text-right group-hover:invisible"
          @click.stop="goTo(message.id)"
        >
          {{ new Date(message.date).toLocaleDateString() }}
        </td>
        <td
          class="invisible group-hover:visible absolute right-0 align-middle pr-2 py-2"
        >
          <!-- <span class="align-middle text-sm"> -->
          <a @click.stop="" href="#" class="trash" title="Delete">
            <ArchiveIcon class="w-5 h-5" />
          </a>
          <a
            @click.stop="removeMessage(message)"
            href="#"
            class="trash"
            title="Delete"
          >
            <DeleteIcon class="w-5 h-5" />
          </a>
          <a @click.stop="" href="#" class="mark" title="Mark as unread">
            <MailIcon class="w-5 h-5" />
          </a>
          <a @click.stop="" href="#" class="snooze" title="Snooze">
            <ClockIcon class="w-5 h-5" />
          </a>
          <!-- </span> -->
        </td>
      </tr>
    </tbody>
  </table>
</template>

<script>
import {mapActions} from 'vuex'
import {init, strip} from '~/shared'

export default {
  name: 'MessageList',
  props: {
    messages: {
      type: Array,
      default: () => [],
    },
  },
  data: () => ({
    isCheckAll: false,
    checkedMessages: [],
  }),
  methods: {
    goTo(id) {
      this.$router.push({path: `/${id}`})
    },
    truncate(str) {
      return `${strip(str).substr(0, 50)}...`
    },
    checkAll() {
      this.isCheckAll = !this.isCheckAll
      this.checkedMessages = []
      if (this.isCheckAll) {
        for (var key in this.messages) {
          this.checkedMessages.push(this.messages[key])
        }
      }
    },
    updateCheckAll() {
      if (this.checkedMessages.length == this.messages.length) {
        this.isCheckAll = true
      } else {
        this.isCheckAll = false
      }
    },
  },
}
</script>
