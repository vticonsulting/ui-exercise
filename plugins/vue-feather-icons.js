import Vue from 'vue'
import {
  AlertOctagonIcon,
  ArchiveIcon,
  ArrowLeftIcon,
  ClockIcon,
  DeleteIcon,
  MailIcon,
  MoreVerticalIcon,
  PlusIcon,
  InboxIcon,
  RefreshCwIcon,
  SettingsIcon,
  TagIcon,
} from 'vue-feather-icons'

Vue.component('AlertIcon', AlertOctagonIcon)
Vue.component('ArchiveIcon', ArchiveIcon)
Vue.component('ArrowLeftIcon', ArrowLeftIcon)
Vue.component('ClockIcon', ClockIcon)
Vue.component('DeleteIcon', DeleteIcon)
Vue.component('MailIcon', MailIcon)
Vue.component('MoreVerticalIcon', MoreVerticalIcon)
Vue.component('PlusIcon', PlusIcon)
Vue.component('InboxIcon', InboxIcon)
Vue.component('RefreshIcon', RefreshCwIcon)
Vue.component('SettingsIcon', SettingsIcon)
Vue.component('TagIcon', TagIcon)
