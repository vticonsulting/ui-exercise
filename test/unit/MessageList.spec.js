import {createLocalVue, shallowMount} from '@vue/test-utils'
import MessageList from '@/components/MessageList'
import '@/plugins/vue-feather-icons'

describe('MessageList', () => {
  it('renders the correct markup', () => {
    const wrapper = shallowMount(MessageList, {
      propsData: {
        messages: [
          {
            id: 1,
            subject: 'Subject',
            sender: 'sender@example.com',
            body: 'Message Body',
            tags: ['tag-1', 'tag-2'],
          },
        ],
      },
      data() {
        return {
          isCheckAll: false,
          checkedMessages: [],
        }
      },
    })
    expect(wrapper).toMatchSnapshot()
  })
})
