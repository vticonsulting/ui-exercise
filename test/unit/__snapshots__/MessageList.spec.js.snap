// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageList renders the correct markup 1`] = `
<table class="w-full">
  <thead>
    <tr class="flex border-b text-2xl">
      <th class="pr-3 md:pl-3 text-left font-normal">
        <input type="checkbox" class="form-checkbox">
      </th>
      <th class="flex items-center text-left font-normal pr-2 flex-grow">
        <div class="flex">
          <a href="#" title="Refresh" class="mr-1">
            <refreshicon-stub size="24" class="w-5 h-5"></refreshicon-stub>
          </a>
        </div>
        <div class="flex items-center space-x-2" style="display: none;">
          <a href="#" title="Archive selected">
            <archiveicon-stub size="24" class="w-5 h-5"></archiveicon-stub>
          </a>
          <a href="#" title="Report spam">
            <alerticon-stub size="24" class="w-5 h-5"></alerticon-stub>
          </a>
          <a title="Delete">
            <deleteicon-stub size="24" class="w-5 h-5"></deleteicon-stub>
          </a>
          <a href="#" title="Mark as read">
            <mailicon-stub size="24" class="w-5 h-5"></mailicon-stub>
          </a>
          <a href="#" title="Snooze">
            <clockicon-stub size="24" class="w-5 h-5"></clockicon-stub>
          </a>
        </div>
        <a href="#" title="More" class="mr-1">
          <moreverticalicon-stub size="24" class="w-5 h-5"></moreverticalicon-stub>
        </a>
      </th>
      <th class="msg__col text-left font-normal">
        <a href="#" title="Settings">
          <settingsicon-stub size="24" class="w-5 h-5"></settingsicon-stub>
        </a>
      </th>
    </tr>
  </thead>
  <tbody>
    <tr class="relative group hover:bg-gray-200 hover:shadow cursor-pointer flex flex-col md:flex-row border-b">
      <td class="md:py-2 md:px-3">
        <input type="checkbox" class="form-checkbox" value="[object Object]">
      </td>
      <td class="md:py-2 pr-6">
        sender@example.com
      </td>
      <td class="md:py-2 pr-1 text-black whitespace-no-wrap">
        <span class="mr-1 p-1 bg-gray-400 rounded text-xs capitalize tag-1">
          tag-1
        </span>
        <span class="mr-1 p-1 bg-gray-400 rounded text-xs capitalize tag-2">
          tag-2
        </span>
      </td>
      <td class="md:py-2 pr-4 text-black font-medium xl:flex-no-grow">
        Subject
      </td>
      <td class="md:py-2 hidden xl:block pr-2 flex-grow">
        Message Body...
      </td>
      <td class="md:py-2 pr-2 text-right group-hover:invisible">
        Invalid Date
      </td>
      <td class="invisible group-hover:visible absolute right-0 align-middle pr-2 py-2">
        <a href="#" title="Delete" class="trash">
          <archiveicon-stub size="24" class="w-5 h-5"></archiveicon-stub>
        </a>
        <a href="#" title="Delete" class="trash">
          <deleteicon-stub size="24" class="w-5 h-5"></deleteicon-stub>
        </a>
        <a href="#" title="Mark as unread" class="mark">
          <mailicon-stub size="24" class="w-5 h-5"></mailicon-stub>
        </a>
        <a href="#" title="Snooze" class="snooze">
          <clockicon-stub size="24" class="w-5 h-5"></clockicon-stub>
        </a>
      </td>
    </tr>
  </tbody>
</table>
`;
